### ISSUE TITLE
Some issues about the gradients of slimNet

### ISSUE THREAD

**Issue by baiyu12345** [changlin31/DS-Net#7]
I'm glad to see such a perfect work. I want to ask how the gradient passes through slim and how to set the parameter( ) in the optimizer.

**Comment by baiyu12345** [comment]
Can someone give an explaination about the code:
`for group in optimizer.param_groups:
                for p in group['params']:
                    if not p.grad is None and torch.sum(torch.abs(p.grad.data)) == 0.0:
                        p.grad = None`

**Comment by changlin31** [comment]
Thanks for your interest!
1. Sorry, I don't get your meaning by 'passes through slim'
2. As in this part of the code: https://github.com/changlin31/DS-Net/blob/2ae9455f7e7c2f3fe14b7e1a395811feabb208e3/train.py#L324-L327
when training the supernet, we all the parameters of `model` in optimizer. When training the gate, we only put the parameters of the gate `model.get_gate()` in optimizer.

3. https://github.com/changlin31/DS-Net/blob/2ae9455f7e7c2f3fe14b7e1a395811feabb208e3/dyn_slim/apis/train_slim.py#L113-L116
This code sets the gradient of all the model parameters that are not used in `forward()` function to `None`, in order to prevent applying weight decay on these parameters. These few lines are very useful when training weight-sharing neural networks.

**Comment by baiyu12345** [comment]
Thank you for your reply, I understand some confusion. And I want to know if ****other parameters**（not in smallest model）** in the biggest model will be updated when the smallest model is training? Do those lines of code ensure that other parameters are not updated?

**Comment by changlin31** [comment]
When training the smallest model only, extra parameters in the biggest model will not participate in forward() and backward(). Thus, their gradients will be `0` and then set to `None` by the aforementioned code. Optimizer will only update parameters with gradients, see these lines in `torch.optim.SGD.step()`:
https://github.com/pytorch/pytorch/blob/d6133b2fe6b863dd49fb21641bd04e24e19ac794/torch/optim/sgd.py#L99-L117
So, these code ensure parameters not used in forward are not updated (by weight decay).

When using inplace distillation with biggest model being the teacher, we need to used `torch.no_grad()` or `torch.Tensor.detach()` to stop the gradient calculation of the biggest model.

**Comment by sunnyxiaohu** [comment]
Hi, Thanks for your perfect work. I was wondering whether this case will happen?
DS-Net/dyn_slim/apis/train_slim.py

Lines 113 to 116 in 2ae9455
` for group in optimizer.param_groups: 
     for p in group['params']: 
         if not p.grad is None and torch.sum(torch.abs(p.grad.data)) == 0.0: 
             p.grad = None `
As in sandwish4 mode, we will always forward `largest` mode, and accumulate the gradients of all modes and make optimizer a step.

Thanks.

**Comment by changlin31** [comment]
Hi @sunnyxiaohu ,

You are right. This lines may not be executed in sandwich sampling scheme. But it is always safer to add these lines when training a weight sharing network. In our case, if we used different BN weight for different paths, these lines would be doing their job.
